## Chapter 3, P3: Fitting & Predicting with GBM Models

For classification tasks using **`GradientBoostingClassifier`**, you have two methods for making predictions, each serving a different purpose:

1. **`.predict(X)`**: This method returns the **final predicted class label** (e.g., 0 or 1, 'Yes' or 'No').  

2. **`.predict_proba(X)`**: This method returns the **probability estimates for each class**. For a binary classification problem, it returns a 2D array where each row contains two values: the probability of the negative class and the probability of the positive class.


## Simple Churn Prediction Example (Binary Classification)

In [7]:
from pydoc import classify_class_attrs
from sklearn.ensemble import GradientBoostingClassifier
import pandas as pd

# Sample classification data
c_data = {
    'Tenure': [2, 48, 12, 1, 36, 24],
    'MonthlySpend': [50, 100, 80, 45, 110, 95],
    'Churn': [1, 0, 0, 1, 0, 1] # 1 for Churn, 0 for No Churn
}
c_df = pd.DataFrame(c_data)

X_c = c_df[['Tenure', 'MonthlySpend']]
y_c = c_df['Churn']

# Instantiate and fit the classifier
gbc = GradientBoostingClassifier(n_estimators=100, random_state=42)
gbc.fit(X_c, y_c)

# New data to predict on
new_customers = pd.DataFrame({'Tenure': [3, 25], 'MonthlySpend': [60, 105]})

# Get class predictions
class_predictions = gbc.predict(new_customers)
print(f"Class Predictions: {class_predictions}\n")

# Get probability predictions
proba_predictions = gbc.predict_proba(new_customers)
print("Probability Predictions (No Churn, Churn):")
print(proba_predictions.round(3))


Class Predictions: [1 1]

Probability Predictions (No Churn, Churn):
[[0.    1.   ]
 [0.431 0.569]]


The output from `.predict_proba()` is often more informative than the class label outputs from `.predict()`. It allows you to **understand the model's confidence** and **set custom thresholds for classification** (e.g. "Churn" may be if probability > 0.7).

* The first new customer (`Tenure`: 3, `MonthlySpend`: 60), the predicted class is `1` (Churn). This is because the probability of Churn (1) is greater than the probability of no churn (0.435).  

* The second customer (`Tenure`: 25, `MonthlySpend`: 105), the prediced class is `1` (Churn). However, its probability of Churn is lower compared the new customer (0.569 vs 1).